File: server/java/westfield.server/westfield.server.runtime/src/main/java/org/freedesktop/westfield/server/Registry.java
package org.freedesktop.westfield.server;


import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class Registry {

    private final Set<RegistryResource> registryResources = new HashSet<>();
    private final Map<Integer, Global>  globals           = new HashMap<>();

    Registry() {}

    void bind(final RegistryResource resource,
              final int name,
              final int id,
              final int version) {
        this.globals.get(name)
                    .bindClient(resource.getClient(),
                                id,
                                version);
    }

    public void register(final Global global) {
        if (this.globals.put(global.hashCode(),
                             global) == null) {
            this.registryResources.forEach(registryResource -> registryResource.global(global.hashCode(),
                                                                                          global.getInterfaceName(),
                                                                                          global.getVersion()));
        }
    }

    public void unregister(final Global global) {
        if (this.globals.remove(global.hashCode()) != null) {
            this.registryResources.forEach(registryResource -> registryResource.global(global.hashCode(),
                                                                                          global.getInterfaceName(),
                                                                                          global.getVersion()));
        }
    }

    void publishGlobals(final RegistryResource registryResource) {
        this.globals.entrySet()
                    .forEach(entry -> {
                        final int    name   = entry.getKey();
                        final Global global = entry.getValue();
                        registryResource.global(name,
                                                 global.getInterfaceName(),
                                                 global.getVersion());
                    });
    }

    RegistryResource createResource(final Client client) {
        final RegistryResource registryResource = new RegistryResource(client,
                                                                          1,
                                                                          this);
        this.registryResources.add(registryResource);
        return registryResource;
    }
}

File: server/web/runtime/src/Registry.js
/*
MIT License

Copyright (c) 2017 Erik De Rijcke

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

'use strict'

import RegistryResource from './RegistryResource'
import RegistryRequests from './RegistryRequests'
import Global from './Global'

/**
 * @implements {RegistryRequests}
 */
class Registry extends RegistryRequests {
  constructor () {
    super()
    /**
     * @type {Array<RegistryResource>}
     * @private
     */
    this._registryResources = []
    /**
     * @type {Object.<number,Global>}
     * @private
     */
    this._globals = {}
    /**
     * @type {number}
     * @private
     */
    this._nextGlobalName = 0xffff0000
  }

  /**
   * Register a global to make it available to clients.
   *
   * @param {Object} implementation
   * @param {string}interface_
   * @param {number}version
   * @param {function(Client,number,number):void}bindCallback callback with Client, id, version arguments
   * @return {Global}
   */
  createGlobal (implementation, interface_, version, bindCallback) {
    const name = ++this._nextGlobalName
    const global = new Global(this, implementation, interface_, version, name, bindCallback)
    this._globals[name] = global
    this._registryResources.forEach(registryResource => registryResource.global(global.name, global.interface_, global.version))
    return global
  }

  /**
   * Unregister a global and revoke it from clients.
   *
   * @param {Global} global
   */
  destroyGlobal (global) {
    if (this._globals[global.name]) {
      this._registryResources.forEach(registryResource => registryResource.globalRemove(global.name))
      setTimeout(()=>{
        delete this._globals[global.name]
      },5000)
    }
  }

  /**
   * @param {RegistryResource} registryResource
   */
  publishGlobals (registryResource) {
    Object.entries(this._globals).forEach(([name, global]) => registryResource.global(Number.parseInt(name), global.interface_, global.version))
  }

  /**
   *
   * @param {Client} client
   * @param {number} id
   */
  createRegistryResource (client, id) {
    const registryResource = new RegistryResource(client, id, 1)
    registryResource.implementation = this
    this._registryResources.push(registryResource)
    return registryResource
  }

  /**
   * Binds a new, client-created object to the server using the
   * specified name as the identifier.
   * @param {Client}client
   * @param {RegistryResource}resource
   * @param {number}name unique numeric name of the object
   * @param {string}interface_
   * @param {number}version
   * @param {number}id bounded object
   * @override
   */
  bind (client, resource, name, interface_, version, id) {
    this._globals[name].bindClient(client, id, version)
  }
}

export default Registry
